### PR TITLE
[web3.js] Replace tweetnacl impl

### LINKED ISSUE #26933: web3.js: Investigate replacement for `tweetnacl`
# Overview

This is the ed25519 library that we use to:

* generate keypairs
* sign and verify messages

Properties:

* No dependencies
* Audited
* 10.5K gzipped
* side effects in module factory
* not tree-shakeable
* compatible with React Native
 
image

# Alternatives (ruled out)

## @noble/ed25519

Properties:

* No dependencies
* Audited
* 5.4kb gzipped
* side effects in module factory
* not tree-shakeable because of side effects in module factory
* not compatible with React Native because of dependency on `SubtleCrypto`
* not compatible with current API because `sign()` returns a `Promise`

image

## dazoe/ed25519

Does not run in browser.

## supercop.js

Is >130KB

## nazar-pc/supercop.wasm

Introduces asynchronicity because of WASM initialization step.

## nkeys.js

Just a wrapper around `tweetnacl`.

## substack/ed25519-supercop

Does not run in browser.

## xwi88/crypto

Just a wrapper around `tweetnacl`.

## BENMFeng/Ed25519_DS

Just a wrapper around `tweetnacl`.

## skorotkiewicz/ed25519-keys

Is >45KB

### PR DESCRIPTION
**Pull request by steveluscher**
#### Problem

This PR aims to replace the ed25519 keygen, signing, and verifying algorithms with smaller ones having fewer dependencies.

#### Summary of Changes

* Replaced `tweetnacl` with `@noble/ed25519`

#### Bundle size

Used `package-build-stats` to analyze the size change locally.

```
const {getPackageStats} = require('package-build-stats');
getPackageStats('~/web3.js').then(s => console.log(s));
```

|Before|After|
|---|---|
|https://gist.github.com/steveluscher/3cc75a21d94be656071a5983acd4fae9|https://gist.github.com/steveluscher/a8722c7ad7726d897b625e971492a086|

tl;dr a ~4% reduction in bundle size after gzip.

#### Dependency tree

Replaced `tweetnacl` for `@noble/ed25519`, dropping about 4K gzipped.

|Before|After|
|---|---|
|image|image|

Produced using https://npmgraph.js.org/

#### Performance

Consider this keypair generation code:

```
// perf.js
const {Keypair} = require('@solana/web3.js/lib/index.cjs');

for (let ii = 0; ii < 10000; ii++) {
  Keypair.generate();
}
```

Take 10 trials:

```
 % time node perf.js
```

Results

image

No change.

---

Consider this transaction signing code:

```
const {
  Keypair,
  PublicKey,
  Transaction,
  TransactionInstruction,
} = require('@solana/web3.js/lib/index.cjs');

const keypair = Keypair.generate();
const tx = new Transaction({
  feePayer: keypair.publicKey,
  recentBlockhash: '3CYc7JXH2iEgiyiSFakXMuBJ43ss8ciJzErUfD7jBCuk',
}).add(
  new TransactionInstruction({
    data: Buffer.from('hello world'),
    keys: [],
    programId: new PublicKey('MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr'),
  }),
);

for (let ii = 0; ii < 1000; ii++) {
  tx.sign(keypair);
}
```

Take 10 trials:

```
 % time node perf.js
```

Results:

image

~6% slower.

#### Compatibility notes

Together, this, #27390, and #27127 make `@solana/web3.js` incompatible with React Native because of their use of:

* `BigInt`
* Exponentiation with `BigInt`
* `bigint` literals

Because this team is small, we're likely to recommend that folks upgrade their React Native apps to 0.70 _and_ convert them to run using the Hermes JavaScript engine, which in that version supports _all_ of the above. Support for big integers is critical to most crypto apps, and React Native has now made Hermes the default engine. We would be best advised to skate where the puck is going.

Addresses #26904
Fixes #26933.

### CODE REVIEW

**Comment by paulmillr** [comment]
@steveluscher it seems like the performance bottleneck is somewhere outside of ed25519, since tweetnacl should be 4x slower.
